CARDS-2176: Switch from material-table to material-react-table

veronikaslc opened the pull request:
Affected components:
- New form/subect dialogs
- Subject selector dialogs
- Patient chart
- Users and groups from Admin
- Vocabularies page from Admin
- Variant Files Container

Resources, that helped with examples:
https://www.material-react-table.com/
https://www.material-react-table.dev/?path=/story/prop-playground--default
https://github.com/KevinVandy/material-react-table/tree/main/apps/material-react-table-storybook/stories

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table

marta- commented:
Dashboard -> (+) -> "New..." dialog: huge gap before the list of options needs to be removed
![image](https://user-images.githubusercontent.com/651980/231484751-28557cac-f5a4-4528-909a-300d4b3066bf.png)

Also, warnings in the js console:
* `Warning: Failed prop type: Invalid prop children supplied to ForwardRef(Box), expected a ReactNode.`
* `Warning: Functions are not valid as a React child. This may happen if you return a Component instead of <Component /> from render. Or maybe you meant to call this function rather than return it.`
* `Warning: Invalid value for prop `title` on <div> tag. Either remove it from the element, or pass a string or number value to keep it in the DOM. For details, see https://reactjs.org/link/attribute-behavior `

----


In the "New" dialog select "Questionnaire" -> search for "info" -> only 2 questionnaires are now shown; click Cancel and start over : (+) -> Questionnaire -> the search string is preserved and still only 2 questionnaires are shown. I find this confusing and suggest the search string should be cleared.

----

In the new form dialog, select "Patient information" -> click "Continue" -> in the "Select a subject" dialog click "Create new subject" -> the header and the selected row look too much alike. Revert header back to white or transparent background. Ideally, in these tables where we have just one column, the column label should be displayed in line with the "Search" box to save some vertical space.

![image](https://user-images.githubusercontent.com/651980/231496162-27d6f8f8-6bb1-457c-ba0c-0ea4290bf913.png)

----

**Regression:** Create a Patient with a Patient information form and one without. Then go through the process of creating a Patient information form starting from the dashboard. At the "Select a subject" step, there is no longer a visual distinction between subjects that can be selected and those that cannot because they reached the maxPerSubject for that questionnaire. In the screenshot, `w` is available and `b` isn't. `b` should have a disabled appearance with lighter gray text.

![image](https://user-images.githubusercontent.com/651980/231495657-c7201488-8952-4d89-9979-692aff4e9e7f.png)

----

In the new form dialog, subject selector dialog, and new subject dialog, the user is expected to click on a row to select it but the mouse cursor is default or text instead of pointer so it doesn't properly indicate that the row is clickable. Cursor should be `pointer` on rows.

----
After creating some forms, navigate to the patient chart. All form listings now have unnecessary headings, paddings, and drop shadows making the chart look heavy and misaligned.

![image](https://user-images.githubusercontent.com/651980/231492887-c6c78e03-6937-4935-bf0c-d0fc8b60f1b5.png)

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Fixed regression for Dashboard -> (+) -> New... dialog: huge gap and font

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Clear the search for New Form dialog

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Better style for subject select dialog;
Reset questionnaire selector dialog on back button click
Removed Identifier table header from subject selector

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Fixed regression for a visual distinction between subjects that can be selected and those that cannot because they reached the maxPerSubject for that questionnaire
Set cursor to pointer for row selection

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Fixed styles in the patient chart

veronikaslc commented:
- Fixed regressions and styles
- Console error seems to be produced by the material table components rendering (afai tested)

marta- commented:
**Create new subject** dialog:
* small issue: "Select a type" looked much better beside the search box. I see MaterialReactTable does not have a way to specify a "title", but maybe we can insert a title in its toolbar?
* bigger issue: the pagination footer reports 2 rows when in reality e only have one subject type

![image](https://user-images.githubusercontent.com/651980/231765218-fe15ed63-6c41-4be5-a07b-f13c672e6eed.png)

----

**Patient chart**: the headers and drop shadows are gone, but the chart is still very misaligned.

![image](https://user-images.githubusercontent.com/651980/231766361-fb8ee299-4151-455a-9644-41ae2dd86a54.png)

----

**Users** and **Groups**: I find the toolbar unnecessary, can we just have the expanded search instead?

![image](https://user-images.githubusercontent.com/651980/231775478-0745b101-3e4c-4832-b9ea-272518401d61.png)

Also, in **Groups**, the columns "Members" and "Declared members" are too wide for displaying just numbers, please reduce the size.

![image](https://user-images.githubusercontent.com/651980/231776312-348acafa-2290-42df-ae1d-c632bd5d072e.png)

marta- commented:
**Bug**:
- create 2 patients
- go to the "Patients" page (from the sidebar link)
- click (+) to open subject creation dialog
- enter an id and select the subject type "Visit", click continue
=> **the parent selection dialog has empty rows**:
![image](https://user-images.githubusercontent.com/651980/231778098-a7035204-9949-48f5-b642-ead08e48ba13.png)

marta- commented:
> * Console error seems to be produced by the material table components rendering (afai tested)

They only pop up for the dashboard's "New..." table, not for any of the other tables, so there should be a way to eliminate them by tweaking that table's settings.

marta- commented:
The variants file versions table is not looking good, all columns are way too wide.

| Now | Before |
|----|----|
|![image](https://user-images.githubusercontent.com/651980/231901372-c61801b7-3f05-4650-9523-d034a4895c0a.png) | ![image](https://user-images.githubusercontent.com/651980/231917725-46f8a8e8-8e2d-4fe5-99ce-560ea35d6027.png) |

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Fixed Select subject type dialog header;
Removed toolbar from users and groups table
Removed redundant table property
Better spacing for table columns for groups

marta- commented:
**Bug:** Administration -> Groups -> (+) -> enter a group name, click "Create group" => the group gets created, but the creation dialog doesn't close, the Groups table doesn't get updated, and there's a js console error:
```Uncaught (in promise) TypeError: Cannot read properties of null (reading 'resetRowSelection')```

marta- commented:
**Bug**: start `lfs` -> create some patients -> from the Dashboard, start creating a Tumor => at the parent selection step, the available patients don't load and there's a js console error in `SubjectSelector`:

```Uncaught (in promise) ReferenceError: changeType is not defined```

![image](https://user-images.githubusercontent.com/651980/232055310-21310c62-3b57-4fbd-8522-fdc8db52eb9f.png)

marta- commented:
And some good news: with the MaterialTable, the tooltip of the pagination buttons was always obstructing the form buttons in New form Dialog and SubjectSelector. With MaterialReactTable this problem is gone and the user experience in these dialogs is so much better!

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Better spacing for variants file versions table;
Fixed bug with selecting parent subject table;
Fixed bug with group creation;

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Fixed bug wuth number of rows in the new subject dialog

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -576,28 +616,28 @@ function SubjectMemberInternal (props) {
+                positionActionsColumn="last"
+                renderRowActions={({ row }) => (
+                    <Box sx={{ display: 'flex', flexWrap: 'nowrap', gap: '8', float: 'right' }}>
+                    <>
```
Is this `<>` necessary ?

marta- commented on modules/homepage/src/main/frontend/src/themePage/index.jsx:
```diff
@@ -33,7 +33,7 @@ import PageStart from "../PageStart";
 import DialogueLoginContainer, { GlobalLoginContext } from "../login/loginDialogue.js";
 
-// Temporary fix for the duplicate displayedRows occurring in material-table 2.0.3
+// Temporary fix for the duplicate displayedRows occurring in material-react-table 2.0.3
```
This temporary fix probably no longer applies, please check and remove

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Aligned subject page table

veronikaslc commented:
All comments addressed except console warnings - does not seem to find the cause.

sdumitriu reviewed (commented):
> Uncaught TypeError: row.original.tableData is undefined
    Cell Subject.jsx:608

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -517,57 +518,115 @@ function SubjectMemberInternal (props) {
-                      fontWeight: "bold",
-                      paddingTop: "10px",
-                      whiteSpace: 'nowrap',
+                muiTableHeadProps={{
+                  sx: {
+                    display: 'none',
+                  },
+                }}
+                muiTableFooterProps={{
+                  sx: {
+                    display: 'none',
+                  },
+                }}
```
can we replace this styling with `enableTableHead={false} enableTableFooter={false}` ? Same comment for all other tables where we hide head and footer with `sx`.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -545,62 +545,81 @@ function SubjectMemberInternal (props) {
+                    size: 400,
                     muiTableBodyCellProps: ({ cell }) => ({
-                      sx: {
+                      sx: (theme) => ({
```
`theme` is not used.

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Fixed Uncaught TypeError: row.original.tableData is undefined

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Removed unnecessary elements

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Removed outdated comment
Replaced styling with corresponding table props
Removed unnecesary React element
Fixed console warning

veronikaslc commented:
Addressed comments and also fixed console warning for the dashboard "New.." dialog open action.

marta- commented on modules/homepage/src/main/frontend/src/themePage/index.jsx:
```diff
@@ -33,7 +33,6 @@ import PageStart from "../PageStart";
 import IndexStyle from "./indexStyle.jsx";
 import DialogueLoginContainer, { GlobalLoginContext } from "../login/loginDialogue.js";
 
-// Temporary fix for the duplicate displayedRows occurring in material-table 2.0.3
```
I meant remove the `materialTableStyles` completely (the definition here and the usage on line 114), not only remove the comment.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -517,57 +518,107 @@ function SubjectMemberInternal (props) {
+                layoutMode="grid"
+                muiTableHeadCellProps={{
+                    sx: {
+                      flex: '0 0 auto',
```
Do we still need this if `enableTableHead={false}`?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -517,57 +518,107 @@ function SubjectMemberInternal (props) {
-                    render: rowData => <React.Fragment>
-                                         {rowData["statusFlags"].map((status) => {
+                }}
+                renderDetailPanel={({ row }) => <FormData formID={row.original["@name"]} maxDisplayed={maxDisplayed} classes={classes}/> }
```
We should take the opportunity of this refactoring to align the excerpt better with the form info row:
| Now | Preferred |
| ---- | ---- |
| ![image](https://user-images.githubusercontent.com/651980/233126773-9bd46e83-5780-47ed-8523-8f7540deffaa.png) | ![image](https://user-images.githubusercontent.com/651980/233126121-7570af14-8850-4e49-957c-7bb79ed001a1.png) |

(note the left alignment, reduced vertical spacing, and absence of border between the expanded row and the detail panel)

veronikaslc replied:
we can not selectively turn off border between the expanded row and the detail panel and not turn it off elsewhere

marta- replied:
Too bad. I guess we'll keep that border.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -517,57 +518,107 @@ function SubjectMemberInternal (props) {
+                      },
+                    }),
+                    Cell: ({ renderedCellValue, row }) => (
+                                   <Grid container direction="row" spacing={1} justifyContent="flex-start">
```
This was a great idea and the alignment is so much better than before. On a narrower screen, the long title "UHN Patient Experience Emergency Department Survey" is sliding under the avatar, so let's put `nowrap` on the grid container:

![image](https://user-images.githubusercontent.com/651980/233135606-621e638c-c3db-43ad-beee-5ce0ae6ae2ae.png)

marta- replied:
We should explore the feasibility of https://phenotips.atlassian.net/browse/CARDS-1322 with MaterialReactTable. **It's out of scope for this refactoring** (unless super easy to do), but if this kind of grouping is possible it will add great flexibility to data access on the subject chart.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -576,28 +627,26 @@ function SubjectMemberInternal (props) {
+                enableRowActions
+                positionActionsColumn="last"
+                renderRowActions={({ row }) => (
+                    <Box sx={{ display: 'flex', flexWrap: 'nowrap', gap: '8', float: 'right' }}>
```
In `sx`, `gap` needs to be specified with a unit of measurement (e.g. `8px`), a number alone is invalid. So `gap: '8'` has no effect here. Which is fine because it looks good without a gap, so it can just be reoved.
```suggestion
                    <Box sx={{ display: 'flex', flexWrap: 'nowrap', float: 'right' }}>
```

The styling needs to be adjusted on the actions cell though, because the delete buttons are not only misaligned with the top menu, but also look cut off on hover:

![image](https://user-images.githubusercontent.com/651980/233133419-885562c2-d2ae-4e0a-b8a8-17345324bd86.png)

I believe we need a 16px right padding on the cell.

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Various style adjustments

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Using theme.spacing

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -576,28 +628,26 @@ function SubjectMemberInternal (props) {
+                enableRowActions
+                positionActionsColumn="last"
+                renderRowActions={({ row }) => (
+                    <Box sx={{ display: 'flex', flexWrap: 'nowrap', float: 'right' }}>
```
I just noticed the `textAlign: right` on the action cells, i don't think we need `float: right` here.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -276,99 +354,51 @@ function NewFormDialog(props) {
                     // /* It doesn't seem possible to alter the className from here */
-                    backgroundColor: (selectedQuestionnaire?.["jcr:uuid"] === rowData["jcr:uuid"]) ? theme.palette.grey["200"] : theme.palette.background.default,
+                    backgroundColor: (selectedQuestionnaire?.["jcr:uuid"] === cell.row.original["jcr:uuid"]) ? theme.palette.grey["200"] : theme.palette.background.default,
                     // // grey out subjects that have already reached maxPerSubject
-                    color: ((relatedForms?.length && (selectedSubject || currentSubject) && (relatedForms.filter((i) => (i["@name"] == rowData["@name"])).length >= (+(rowData?.["maxPerSubject"]) || undefined)))
+                    color: ((relatedForms?.length && (selectedSubject || currentSubject) && (relatedForms.filter((i) => (i["@name"] == cell.row.original["@name"])).length >= (+(cell.row.original?.["maxPerSubject"]) || undefined)))
```
These custom colors have always bothered me. Can we leverage MaterialReactTable's row selection functionality https://www.material-react-table.com/docs/guides/row-selection? I'm not sure whether manual selection (to preserve current ui) or radio buttons are better. I have a feeling that adding radio buttons to the NewFormDialog, SubjectSelector, etc tables will actually improve user experience, as it will be more clear what they need to do on that screen.

marta- replied:
Based on a quick poll, let's implement manual single selection on click (no radio buttons). The table component should take care of the "selected" styling.

veronikaslc replied:
Addressed in [6bd974e](https://github.com/data-team-uhn/cards/pull/1452/commits/6bd974e851ea5ce1a9c1bcdba4da583fd4afb9e9)

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Fixed pagination row count for subject selection
Removed redundant properties
Use built-in row selection for background change of selected rows
Better alignment

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
A little material react table version bump to the latest 1.10.0

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -462,57 +463,102 @@ function SubjectMemberInternal (props) {
+                muiTableBodyCellProps={{
+                  sx: {
+                    flex: '0 0 auto',
+                  },
+                }}
+                muiTableDetailPanelProps={{
+                  sx: {
+                    paddingTop: '0',
```
With the border still there between above the details panel, it looks better if we do not remove the top padding.
```suggestion
```

marta- commented on lfs-resources/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -911,48 +911,64 @@ export default function VariantFilesContainer() {
+            { header: 'Created', size: 10,
+              muiTableBodyCellProps: {
+                sx: {
+                  paddingLeft: 0,
```
At least the first cell should have some left padding, it looks a bit awkward otherwise:
![image](https://user-images.githubusercontent.com/651980/233679452-0d9676eb-2eaf-44dd-9aa5-6fa846e5564e.png)

marta- reviewed (commented):
can we make the line under the details panel go all the way to the right?
![image](https://user-images.githubusercontent.com/651980/233678063-d636ab67-1dff-4f5e-9a59-6d94dc91b1e2.png)

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table

Added back the top padding for the details panel in patient chart

Co-authored-by: Marta Girdea <marta@phenotips.org>

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Added space to tle left of the variant file names table

marta- commented on lfs-resources/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -911,48 +911,64 @@ export default function VariantFilesContainer() {
+            { header: 'Created', size: 10,
+              muiTableBodyCellProps: {
+                sx: (theme) => ({
+                  paddingLeft: theme.spacing(1),
```
Still not properly aligned:
![image](https://user-images.githubusercontent.com/651980/233742005-aa63d3ba-0a5f-46af-b53d-4d04dc04f659.png)
```suggestion
                  paddingLeft: theme.spacing(2),
```

marta- committed:
CARDS-2176: Switch from material-table to material-react-table

Better alignment in the Variants file versions table

marta- committed:
CARDS-2176: Switch from material-table to material-react-table

Make the Variants file versions dialog look less cluttered

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Made the line under the details panel go all the way to the right

marta- reviewed (approved):
Looks good to me now. But some more testing and a pair of fresh eyes would be great.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -276,99 +354,51 @@ function NewFormDialog(props) {
+                  sx: {
+                    // grey out subjects that have already reached maxPerSubject
+                    color: ((relatedForms?.length && (selectedSubject || currentSubject) && (relatedForms.filter((i) => (i["@name"] == cell.row.original["@name"])).length >= (+(cell.row.original?.["maxPerSubject"]) || undefined)))
                     ? theme.palette.grey["500"]
                     : theme.palette.grey["900"]
```
These should be replaced by 
```suggestion
                    ? theme.palette.text.disabled
                    : theme.palette.text.primary
```

marta- reviewed (dismissed):
New form dialog: rows that are technically disabled should **not** look selected once clicked:
![image](https://user-images.githubusercontent.com/651980/234101550-feb11b64-1a09-4781-9433-aeec83071692.png)

New form dialog, Subject selector,  dashboard's "New..." dialog: the hover background for selected rows is identical to the hover background for unselected rows, so when clicking on a row it doesn't look like the click had any effect. In the Material React Table row selection examples, selected rows have a different background color on hover - update the code to ensure the same happens here.

marta- commented:
> New form dialog, Subject selector, dashboard's "New..." dialog: the hover background for selected rows is identical to the hover background for unselected rows, so when clicking on a row it doesn't look like the click had any effect.

After some more investigation, I believe this is a side-effect of #1447 that we did not notice when merging. It will still have to be fixed, but it's probably not going to be something to do with MaterialReactTable itself.

marta- committed:
CARDS-2176: Switch from material-table to material-react-table
- Refactored some of the old NewFormDialog code to reduce code duplication
- Tweaked the styling to make better use of theme colors

marta- commented:
> New form dialog: rows that are technically disabled should **not** look selected once clicked

> These should be replaced by ...

Addressed in c143d70

sdumitriu committed:
CARDS-2176: Switch from material-table to material-react-table
Bugfix: trying to add a user to a group throws exception due to missing import

sdumitriu reviewed (commented):
- No more autofocus on the subject/questionnaire search input when creating a new form?
- prems mode, create a new patient information form, go to the subject page, +, click on Canadian Patient Experience Survey -> appears selected; click on Patient Information -> error message displayed, but CPES still appears to be selected, however the Continue button is disabled; either deselect CPES, or keep it actually selected and let the Continue button work
- The "loading" placeholder for the table has an out-of-place blue background (can be seen if throttling the connection when going to the form creation dialogs). Can that color be adjusted (white), or the loading placeholder disabled?
- There was a missing import which I fixed in the last commit

veronikaslc commented:
>No more autofocus on the subject/questionnaire search input when creating a new form?
React material table do not have this setting. We can try something like `document.querySelector('div input[type=text][placeholder=Search]').focus()` inside `useEffect()`, but it may be too hackish. 

>prems mode, create a new patient information form, go to the subject page, +, click on Canadian Patient Experience Survey..
Can not follow the path, where to click on subject page to get Canadian Patient Experience Survey or where to go after + to get this servey?

>The "loading" placeholder for the table has an out-of-place blue background..
Which one in which table?

![image](https://user-images.githubusercontent.com/4371097/237042758-ff4b5633-1b54-477b-ad17-54dea8dc814f.png)

sdumitriu commented:
Placeholder:

![table-plaheholder](https://github.com/data-team-uhn/cards/assets/88663/f030458f-9869-48d9-919d-64eb571bf6d1)

Normal:

![table-normal](https://github.com/data-team-uhn/cards/assets/88663/c2c81c11-742a-4de5-b2c4-21c5fc5fbecf)


The blue is not actually entirely new, it's the "row selected" color:

![table-selected](https://github.com/data-team-uhn/cards/assets/88663/ea2a7a2c-94cf-44da-b842-ee2862ea6a02)

sdumitriu commented:
prems mode, create a new patient information form, go to the subject page, +

![step1](https://github.com/data-team-uhn/cards/assets/88663/8c9b64ef-a194-4e22-902d-12c1fd88e801)

Click on Canadian Patient Experience Survey -> appears selected

![step2](https://github.com/data-team-uhn/cards/assets/88663/2fd721e4-b885-4376-a6e6-ae8cde6ef714)

click on Patient Information -> error message displayed, but CPES still appears to be selected

![step3](https://github.com/data-team-uhn/cards/assets/88663/2336f7ae-9767-4c58-a330-3ae440c1d4d7)

However the Continue button is disabled; either deselect CPES, or keep it actually selected and let the Continue button work

sdumitriu commented:
> However the Continue button is disabled

Clarification: the button does not **appear** disabled, it is still blue and clickable, but clicking it does nothing because in the JS code ["progress" is disabled](https://github.com/data-team-uhn/cards/blob/3db79590338c989905ed16328402acfa84a37f0e/modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx#L939).

sdumitriu commented:
> React material table does not have this setting.

Can you try passing something like `muiSearchTextFieldProps={{ autoFocus: true }}`?

> We can try something like document.querySelector('div input[type=text][placeholder=Search]').focus() inside useEffect(), but it may be too hackish.

Indeed, too ugly, let's not do that.

sdumitriu commented:
I think the loading placeholder can be styled with `muiTableBodyCellSkeletonProps`, or the classname `.MuiSkeleton-root`.

marta- commented:
> Placeholder:

I like the placeholder very much, I would like to keep it

sdumitriu commented:
> > Placeholder:
> 
> I like the placeholder very much, I would like to keep it

Yes, but do you mind if it's white instead of blue?

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
Added autofocus for new subject dialogs

veronikaslc commented:
Fixed autofocus issue

marta- commented:
> > > Placeholder:
> > 
> > 
> > I like the placeholder very much, I would like to keep it
> 
> Yes, but do you mind if it's white instead of blue?

Ok

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table
White background of the loading subject table rows

veronikaslc commented:
> I think the loading placeholder can be styled with muiTableBodyCellSkeletonProps, or the classname .MuiSkeleton-root.

Both did not work, loading props are controlled internally by the setting `selected` status on rows.

veronikaslc commented:
> > > > Placeholder:
> > > 
> > > 
> > > I like the placeholder very much, I would like to keep it
> > 
> > 
> > Yes, but do you mind if it's white instead of blue?
> 
> Ok

Fixed

sdumitriu reviewed (approved):
Looks good.

I can still reproduce this bug on all my browsers and computers though:

> prems mode, create a new patient information form, go to the subject page, +, click on Canadian Patient Experience Survey -> appears selected; click on Patient Information -> error message displayed, but CPES still appears to be selected, however the Continue button is disabled; either deselect CPES, or keep it actually selected and let the Continue button work

marta- commented:
> prems mode, create a new patient information form, go to the subject page, +, click on Canadian Patient Experience Survey -> appears selected; click on Patient Information -> error message displayed, but CPES still appears to be selected, however the Continue button is disabled; either deselect CPES, or keep it actually selected and let the Continue button work

@sashaandjic can you try to replicate @sdumitriu's issue ?

veronikaslc commented:
Still can not reproduce this issue. The build starts with no errors, but after the steps `go to the subject page, +,` i only see this:
![image](https://github.com/data-team-uhn/cards/assets/4371097/310fcfc7-4161-4f52-87cd-d9ca0d7b1837)

sdumitriu commented:
I know why the wrong color is applied on hover. The styling for hovered cells is dynamically computed based on the row's "hover" and "selected" state:
- if the row is hovered and selected, use the theme's primary color
- if the row is hovered and not selected, use the theme's background color
- if the row is not hovered, don't apply any special background

However, the way we handle row selection is not the proper way.

Basically, there are two ways of handling row selection: let MRT do it, or do it in a custom way with our our onClick handler. We're doing the latter, but for styling purposes, MRT only looks at its internal "is row selected" state. So we are artificially coloring the row as selected, and **we** know the row is selected, but MRT doesn't know that. In theory, there is the `row.toggleSelected()` call, but looking at the sources for MRT, that doesn't do anything. We should let MRT handle row selection by following [this guide](https://www.material-react-table.com/docs/guides/row-selection#manual-row-selection-without-checkboxes).

I just made it work for the first dialog, selecting what to be created, will push a commit in a few minutes.

sdumitriu commented:
> Still can not reproduce this issue. The build starts with no errors, but after the steps `go to the subject page, +,` i only see this:

You or not on the subject page, you are on the "Subjects list" page. The subject page is a specific patient chart.

sdumitriu reviewed (commented):
Leftover code causing a crash: one instance of `onQueryChange` in `SubjectSelector`.

sdumitriu committed:
CARDS-2176: Switch from material-table to material-react-table
Let MRT know which row is selected

sdumitriu committed:
CARDS-2176: Switch from material-table to material-react-table
Don't allow selecting an invalid subject, just warn about it

sdumitriu committed:
CARDS-2176: Switch from material-table to material-react-table
Better selection of the text color in the subject selector list

sdumitriu committed:
CARDS-2176: Switch from material-table to material-react-table
Regression: Wrong questionnaire list is shown when creating a new form from a subject chart

veronikaslc committed:
CARDS-2176: Switch from material-table to material-react-table

Removed the leftover code
